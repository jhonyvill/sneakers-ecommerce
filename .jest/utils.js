import { render } from "@testing-library/react";
import { ThemeProvider } from "styled-components";
import theme from "../src/Themes";

export function renderWithTheme(children) {
  return render(<ThemeProvider theme={theme}>{children}</ThemeProvider>);
}

export const productDefault = {
  id: "id-1",
  brand: "brand mock",
  title: "title mock",
  price: 250,
  discount: 0.5,
  description: "description mock",
  pictures: [
    {
      src: "image-1.jpg",
      thumb: "thumbnail-1.jpg",
      alt: "product 1",
    },
    {
      src: "image-2.jpg",
      thumb: "thumbnail-2.jpg",
      alt: "product 2",
    },
    {
      src: "image-3.jpg",
      thumb: "thumbnail-3.jpg",
      alt: "product 3",
    },
    {
      src: "image-4.jpg",
      thumb: "thumbnail-4.jpg",
      alt: "product 4",
    },
    {
      src: "image-5.jpg",
      thumb: "thumbnail-5.jpg",
      alt: "product 5",
    },
  ],
};

export const itemDefault = {
  product: { ...productDefault },
  quantity: 1,
};
