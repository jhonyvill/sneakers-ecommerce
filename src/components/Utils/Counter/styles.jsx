import styled from "styled-components";

export const Container = styled.div`
  background-color: ${({ theme }) => theme.colors.lightGrayishBlue};
  font-size: ${({ theme }) => theme.sizes.m};
  color: ${({ theme }) => theme.colors.blackOpacity};
  border-radius: 0.625rem;
  font-weight: 700;
  display: flex;
  align-items: center;
  justify-content: space-between;
  /* grid-column: 1; */
  height: 3.5rem;
  width: 15em;
  /* min-width: 100px; */
  
  @media (max-width: ${({theme}) => theme.screen.tablet}){
    width: 180px;

    }
    @media (max-width: ${({theme}) => theme.screen.largeMobile}){
      width: 100%;
    }
`;

export const CounterButton = styled.button`
  background-color: transparent;
  display: flex;
  align-items: center;
  padding: 1rem;
  height: 100%;

  &:hover {
    opacity: 0.7;
  }
  & svg {
    scale: 0.8;
  }
  @media (max-width: ${({ theme }) => theme.screen.largeMobile}) {
    & svg {
      scale: 1;
    }
  }
`;
