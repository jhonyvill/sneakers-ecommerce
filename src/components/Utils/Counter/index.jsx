import * as S from "./styles";
import MinusSvg from "../../../assets/icon-minus.svg?react";
import PlusSvg from "../../../assets/icon-plus.svg?react";

const operator = {
  sum: "+",
  subtraction: "-",
};

const Counter = ({ count, setCount }) => {
  function handleCounter(operation) {
    if (operation === operator.sum)
      count < 10 && setCount((count) => count + 1);
    if (operation === operator.subtraction)
      count > 0 && setCount((count) => count - 1);
  }

  return (
    <S.Container>
      <S.CounterButton onClick={() => handleCounter(operator.subtraction)} aria-label="menos">
        <MinusSvg />
      </S.CounterButton>
      <span>{count}</span>
      <S.CounterButton onClick={() => handleCounter(operator.sum)} aria-label="mais">
        <PlusSvg />
      </S.CounterButton>
    </S.Container>
  );
};

export default Counter;
