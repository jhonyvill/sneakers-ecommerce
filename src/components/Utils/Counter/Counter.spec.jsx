import { screen } from "@testing-library/react";
import userEvent from "@testing-library/user-event";
import { renderWithTheme } from "../../../../.jest/utils";
import Counter from ".";

function renderCounter(count, setCount = jest.fn()) {
  return renderWithTheme(<Counter count={count} setCount={setCount} />);
}

describe("<Counter />", () => {
  it("should render Counter component", () => {
    const { getByTestId } = renderCounter(0);

    expect(screen.getByRole("button", { name: "menos" })).toBeInTheDocument();
    expect(screen.getByRole("button", { name: "mais" })).toBeInTheDocument();
    expect(getByTestId("counter")).toBeInTheDocument();
    expect(getByTestId("counter")).toHaveTextContent("0");
  });

  it("should call setCount in decrement function with 0 when counter is 1", async () => {
    const setCountMock = jest.fn();
    renderCounter(1, setCountMock);;

    const decrementButton = screen.getByRole("button", { name: "menos" });
    await userEvent.click(decrementButton);

    expect.assertions(4);
    expect(decrementButton).toBeInTheDocument();
    expect(setCountMock).toHaveBeenCalledTimes(1);
    expect(setCountMock).toHaveBeenCalledWith(expect.any(Function));
    expect(setCountMock.mock.calls[0][0](1)).toBe(0);
  });
  it("shouldn't call 'setCount' in decrement function when counter is 0", async () => {
    const mockFunction = jest.fn();
    renderCounter(0, mockFunction);

    const decrementButton = screen.getByRole("button", { name: "menos" });
    await userEvent.click(decrementButton);

    expect.assertions(2);
    expect(decrementButton).toBeInTheDocument();
    expect(mockFunction).not.toHaveBeenCalled();
  });

  it("should call setCount in increment function with 1 when counter is 0", async () => {
    const setCountMock = jest.fn();
    renderCounter(0, setCountMock);;

    const incrementButton = screen.getByRole("button", { name: "mais" });
    await userEvent.click(incrementButton);

    expect.assertions(4);
    expect(incrementButton).toBeInTheDocument();
    expect(setCountMock).toHaveBeenCalledTimes(1);
    expect(setCountMock).toHaveBeenCalledWith(expect.any(Function));
    expect(setCountMock.mock.calls[0][0](0)).toBe(1);
  });
  
  it("shouldn't call 'setCount' in increment function when counter is 10", async () => {
    const mockFunction = jest.fn();
    renderCounter(10, mockFunction);

    const incrementButton = screen.getByRole("button", { name: "mais" });
    await userEvent.click(incrementButton);

    expect.assertions(2);
    expect(incrementButton).toBeInTheDocument();
    expect(mockFunction).not.toHaveBeenCalled();
  });  
});

describe("<Counter /> snapshots", () => {
  it("should match snapshot", () => {
    const { container } = renderCounter(0);
    expect(container.firstChild).toMatchSnapshot();
  });
});

