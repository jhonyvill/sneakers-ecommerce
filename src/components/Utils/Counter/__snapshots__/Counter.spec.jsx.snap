// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Counter /> snapshots should match snapshot 1`] = `
<section
  aria-label="selecione a quantidade de itens"
  class="styles__Container-sc-6vyuvp-0 klSogO"
>
  <button
    aria-label="menos"
    class="styles__CounterButton-sc-6vyuvp-1 hNfiSo"
  >
    <icon-mock />
  </button>
  <span
    data-testid="counter"
  >
    0
  </span>
  <button
    aria-label="mais"
    class="styles__CounterButton-sc-6vyuvp-1 hNfiSo"
  >
    <icon-mock />
  </button>
</section>
`;
